src/graph/network.cc

```diff
@@ -78,104 +156,218 @@ DGL_REGISTER_GLOBAL("network._CAPI_DGLSenderConnect")
+    }
+  });
+
+
+////////////////////////// Distributed Sampler Components ////////////////////////////////
+
+
+DGL_REGISTER_GLOBAL("network._CAPI_SenderSendNodeFlow")
 .set_body([] (DGLArgs args, DGLRetValue* rv) {
     CommunicatorHandle chandle = args[0];
     int recv_id = args[1];
-    // TODO(minjie): could simply use NodeFlow nf = args[2];
     GraphRef g = args[2];
-    const IdArray node_mapping = args[3];
-    const IdArray edge_mapping = args[4];
-    const IdArray layer_offsets = args[5];
-    const IdArray flow_offsets = args[6];
+    NDArray node_mapping = args[3];
+    NDArray edge_mapping = args[4];
+    NDArray layer_offsets = args[5];
+    NDArray flow_offsets = args[6];
     auto ptr = std::dynamic_pointer_cast<ImmutableGraph>(g.sptr());
     CHECK(ptr) << "only immutable graph is allowed in send/recv";
-    network::Sender* sender = static_cast<network::Sender*>(chandle);
     auto csr = ptr->GetInCSR();
-    // Write control message
-    char* buffer = sender->GetBuffer();
-    *buffer = CONTROL_NODEFLOW;
-    // Serialize nodeflow to data buffer
-    int64_t data_size = network::SerializeSampledSubgraph(
-                             buffer+sizeof(CONTROL_NODEFLOW),
-                             csr,
-                             node_mapping,
-                             edge_mapping,
-                             layer_offsets,
-                             flow_offsets);
-    CHECK_GT(data_size, 0);
-    data_size += sizeof(CONTROL_NODEFLOW);
-    // Send msg via network
-    SendData(sender, buffer, data_size, recv_id);
+    // Create a message for the meta data of ndarray
+    NDArray indptr = csr->indptr();
+    NDArray indice = csr->indices();
+    NDArray edge_ids = csr->edge_ids();
+    MsgMeta msg(kNodeFlowMsg);
+    msg.AddArray(node_mapping);
+    msg.AddArray(edge_mapping);
+    msg.AddArray(layer_offsets);
+    msg.AddArray(flow_offsets);
+    msg.AddArray(indptr);
+    msg.AddArray(indice);
+    msg.AddArray(edge_ids);
+    // send meta message
+    int64_t size = 0;
+    char* data = msg.Serialize(&size);
+    network::Sender* sender = static_cast<network::Sender*>(chandle);
+    Message send_msg;
+    send_msg.data = data;
+    send_msg.size = size;
+    send_msg.deallocator = DefaultMessageDeleter;
+    CHECK_NE(sender->Send(send_msg, recv_id), -1);
+    // send node_mapping
+    Message node_mapping_msg;
+    node_mapping_msg.data = static_cast<char*>(node_mapping->data);
+    node_mapping_msg.size = node_mapping.GetSize();
+    node_mapping_msg.aux_handler = &node_mapping;
+    node_mapping_msg.deallocator = NDArrayDeleter;
+    CHECK_NE(sender->Send(node_mapping_msg, recv_id), -1);
+    // send edege_mapping
+    Message edge_mapping_msg;
+    edge_mapping_msg.data = static_cast<char*>(edge_mapping->data);
+    edge_mapping_msg.size = edge_mapping.GetSize();
+    edge_mapping_msg.aux_handler = &edge_mapping;
+    edge_mapping_msg.deallocator = NDArrayDeleter;
```

Comment:
Is this  really a right way of deallocating the NDArray? `edge_mapping` is an object in the stack. But `NDArrayDeleter` call `delete` to free the pointer to the object in the stack. It seems to cause serious memory error. Do I miss something here?

Reply:
It is an error and will be fixed in a new small PR.